samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,189 @@
+	manifest *internal.Manifest
+
+	// pollerStarted, if true represents rule and target pollers are started.
+	pollerStarted bool
+
+	// samplingRulesPollingInterval, default is 300 seconds.
+	samplingRulesPollingInterval time.Duration
+
+	// matching attribute
+	serviceName string
+
+	// matching attribute
+	cloudPlatform string
+
+	// fallback sampler
+	fallbackSampler *FallbackSampler
+
+	// logger for logging
+	logger logr.Logger
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ sdktrace.Sampler = (*remoteSampler)(nil)
+
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, serviceName string, cloudPlatform string, opts ...Option) (sdktrace.Sampler, error) {
+	// create new config based on options or set to default values
+	cfg := newConfig(opts...)
+
+	// validate config
+	err := validateConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// create manifest with config
+	m, err := internal.NewManifest(cfg.endpoint, cfg.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteSampler := &remoteSampler{
+		manifest:                     m,
+		samplingRulesPollingInterval: cfg.samplingRulesPollingInterval,
+		fallbackSampler:              NewFallbackSampler(),
+		serviceName:                  serviceName,
+		cloudPlatform:                cloudPlatform,
+		logger:                       cfg.logger,
+	}
+
+	// starts the rule and target poller
+	remoteSampler.start(ctx)
+
+	return remoteSampler, nil
+}
+
+// ShouldSample matches span attributes with retrieved sampling rules and perform sampling,
+// if rules does not match or manifest is expired then use fallback sampling.
+func (rs *remoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	if !rs.manifest.Expired() {
+		// match against known rules
+		r, match, err := rs.manifest.MatchAgainstManifestRules(parameters, rs.serviceName, rs.cloudPlatform)
+		if err != nil {
+			rs.logger.Error(err, "regexp matching error while matching span and resource attributes")
+			return sdktrace.SamplingResult{}
+		}
+
+		if match {
+			// remote sampling based on rule match
+			return r.Sample(parameters, rs.manifest.Clock.Now())
+		}
+	}
+
+	// Use fallback sampler if manifest is expired
+	// or sampling rules does not match against manifest
+	rs.logger.V(5).Info("span attributes does not match to the sampling rules or manifest is expired so using fallback sampling strategy")
+	return rs.fallbackSampler.ShouldSample(parameters)
+}
+
+// Description returns description of the sampler being used.
+func (rs *remoteSampler) Description() string {
+	return "AwsXrayRemoteSampler{" + rs.getDescription() + "}"
+}
+
+func (rs *remoteSampler) getDescription() string {
+	return "remote sampling with AWS X-Ray"
```

Comment:
Any reason not to just inline this in `Description()`?

Reply:
Initially it was inlined only but @anuraaga suggested to use `getDescription()` method probably to have implementation parity with Java implementation. I don't have any preference on this. I'm fine with or without the method.

Reply:
Unless there is logic to be performed here, just inline it.  It can always be changed later if we decide there is something interesting to do to build the description.

Reply:
Ah the suggestion was to make the description itself follow the pattern of the java implementation, not the method name

Reply:
oh okay. I might have misread the comment then. Yeah we should be okay to inline this then.